src/Mod/Spreadsheet/Gui/SheetTableView.cpp

```diff
@@ -73,34 +74,91 @@ bool SheetViewHeader::viewportEvent(QEvent *e) {
     return QHeaderView::viewportEvent(e);
 }
 
+static std::pair<int, int> selectedMinMaxRow(QModelIndexList list)
+{
+    int min = std::numeric_limits<int>::max();
+    int max = 0;
+    for (const auto & item : list) {
+        int row = item.row();
+        min = std::min(row, min);
+        max = std::max(row, max);
+    }
+    return std::pair<int, int>{min, max};
+}
+
+static std::pair<int, int> selectedMinMaxColumns(QModelIndexList list)
+{
+    int min = std::numeric_limits<int>::max();
+    int max = 0;
+    for (const auto & item : list) {
+        int column = item.column();
+        min = std::min(column, min);
+        max = std::max(column, max);
+    }
+    return std::pair<int, int>{min, max};
+}
+
 SheetTableView::SheetTableView(QWidget *parent)
     : QTableView(parent)
     , sheet(0)
 {
-    QAction * insertRows = new QAction(tr("Insert rows"), this);
-    QAction * removeRows = new QAction(tr("Remove rows"), this);
-    QAction * insertColumns = new QAction(tr("Insert columns"), this);
-    QAction * removeColumns = new QAction(tr("Remove columns"), this);
-
     setHorizontalHeader(new SheetViewHeader(this,Qt::Horizontal));
     setVerticalHeader(new SheetViewHeader(this,Qt::Vertical));
     setVerticalScrollMode(QAbstractItemView::ScrollPerPixel);
     setHorizontalScrollMode(QAbstractItemView::ScrollPerPixel);
 
-    horizontalHeader()->addAction(insertColumns);
-    horizontalHeader()->addAction(removeColumns);
-    horizontalHeader()->setContextMenuPolicy(Qt::ActionsContextMenu);
-
-    verticalHeader()->addAction(insertRows);
-    verticalHeader()->addAction(removeRows);
-    verticalHeader()->setContextMenuPolicy(Qt::ActionsContextMenu);
-
-    connect(insertRows, SIGNAL(triggered()), this, SLOT(insertRows()));
-    connect(insertColumns, SIGNAL(triggered()), this, SLOT(insertColumns()));
-    connect(removeRows, SIGNAL(triggered()), this, SLOT(removeRows()));
-    connect(removeColumns, SIGNAL(triggered()), this, SLOT(removeColumns()));
-
-    QAction * cellProperties = new QAction(tr("Properties..."), this);
+    horizontalHeader()->setContextMenuPolicy(Qt::CustomContextMenu);
+    verticalHeader()->setContextMenuPolicy(Qt::CustomContextMenu);
+
+    connect(verticalHeader(), &QWidget::customContextMenuRequested,
+       [this](const QPoint &point){
+            QMenu menu(this);
+            const auto selection = selectionModel()->selectedRows();
+            const auto & [min, max] = selectedMinMaxRow(selection);
+            if (bool isContiguous = max - min == selection.size() - 1) {
+                /*: This is shown in the context menu for the vertical header in a spreadsheet.
+                    The number refers to how many lines are selected and will be inserted. */
+                auto insertBefore = menu.addAction(tr("Insert %n row(s) above", "", selection.size()));
+                connect(insertBefore, SIGNAL(triggered()), this, SLOT(insertRows()));
```

Comment:
Do all of our compilers now support the "new" style `connect` that uses function pointers instead of macros? I don't remember which compiler didn't use to like it.

Reply:
It works with Qt5 + C++11 according to: https://wiki.qt.io/New_Signal_Slot_Syntax